File: zeroday-server/src/main/kotlin/com/zeroday/handler/handlers/KMapHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject

/**
 * Knowledge-map endpoints. Players can view discovered fragments, see
 * available fragments to discover, and "discover" one (paying the
 * resource cost and unlocking the associated command or research hook).
 */
class KMapHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.KMAP_GET,
        RequestTypes.KMAP_DISCOVER,
        RequestTypes.KMAP_FRAGMENTS
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.KMAP_GET -> get(ctx, playerId)
            RequestTypes.KMAP_DISCOVER -> discover(ctx, playerId, payload)
            RequestTypes.KMAP_FRAGMENTS -> fragments(ctx, playerId)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported kmap request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun get(ctx: HandlerContext, playerId: String): HandlerResult {
        val kmap = ctx.services.hackToLearnService.getKnowledgeMap(playerId)
        return HandlerResult.Ok(ResponseTypes.KNOWLEDGE_MAP, mapOf("kmap" to kmap))
    }

    private suspend fun discover(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val fragmentId = try { payload.reqStr("fragment_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Fragment ID required", ErrorCategory.VALIDATION))
        }
        return ctx.services.hackToLearnService.discoverFragment(playerId, fragmentId).fold(
            onSuccess = { fragment ->
                HandlerResult.Ok(
                    type = ResponseTypes.FRAGMENT_DISCOVERED,
                    payload = mapOf(
                        "fragment" to fragment,
                        "message" to "Discovered: ${fragment.name}" +
                            (fragment.unlocksCommand?.let { " - Unlocked command: $it" } ?: "")
                    )
                )
            },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Discovery failed", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun fragments(ctx: HandlerContext, playerId: String): HandlerResult {
        val available = ctx.services.hackToLearnService.getAvailableFragments(playerId)
        return HandlerResult.Ok(ResponseTypes.AVAILABLE_FRAGMENTS, mapOf("fragments" to available))
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/KMapHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class KMapHandler : IHandler
{
    private readonly HackToLearnService _hackToLearnService;

    public string MessageType => "kmap";

    public KMapHandler(HackToLearnService hackToLearnService)
    {
        _hackToLearnService = hackToLearnService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return new MessageResult("error", connectionId, new { message = "Missing payload" });

        var action = payload.Value.Str("action");
        return action switch
        {
            "get" => await HandleGet(connectionId, payload.Value),
            "discover" => await HandleDiscover(connectionId, payload.Value),
            "fragments" => await HandleFragments(connectionId, payload.Value),
            _ => new MessageResult("error", connectionId, new { message = $"Unknown kmap action: {action}" })
        };
    }

    private async Task<IActionResult> HandleGet(string connectionId, JsonElement payload)
    {
        var playerId = payload.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var kmap = await _hackToLearnService.GetKnowledgeMap(playerId);

        return new MessageResult("knowledge_map", connectionId, new { kmap });
    }

    private async Task<IActionResult> HandleDiscover(string connectionId, JsonElement payload)
    {
        var playerId = payload.Str("playerId");
        var fragmentId = payload.Str("fragment_id");

        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(fragmentId))
            return new MessageResult("error", connectionId, new { message = "Fragment ID required" });

        try
        {
            var fragment = await _hackToLearnService.DiscoverFragment(playerId, fragmentId);
            var msg = $"Discovered: {fragment.Name}";
            if (fragment.UnlocksCommand is not null)
                msg += $" - Unlocked command: {fragment.UnlocksCommand}";

            return new MessageResult("fragment_discovered", connectionId, new
            {
                fragment,
                message = msg
            });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    private async Task<IActionResult> HandleFragments(string connectionId, JsonElement payload)
    {
        var playerId = payload.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var fragments = await _hackToLearnService.GetAvailableFragments(playerId);

        return new MessageResult("available_fragments", connectionId, new { fragments });
    }
}
